# Making Recommendations Based on Popularity

These datasets are hosted on: https://archive.ics.uci.edu/ml/datasets/Restaurant+%26+consumer+data

They were originally published by: Blanca Vargas-Govea, Juan Gabriel González-Serna, Rafael Ponce-Medellín. Effects of relevant contextual features in the performance of a restaurant recommender system. In RecSys11: Workshop on Context Aware Recommender Systems (CARS-2011), Chicago, IL, USA, October 23, 2011.

## Restaurants data

In [1]:
import numpy as np
import pandas as pd

In [2]:
# rating_final.csv
url = 'https://drive.google.com/file/d/1ptu4AlEXO4qQ8GytxKHoeuS1y4l_zWkC/view?usp=sharing' 
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
frame = pd.read_csv(path)

# chefmozcuisine.csv
url = 'https://drive.google.com/file/d/1S0_EGSRERIkSKW4D8xHPGZMqvlhuUzp1/view?usp=sharing' 
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
cuisine = pd.read_csv(path)

# 'geoplaces2.csv'
url = 'https://drive.google.com/file/d/1ee3ib7LqGsMUksY68SD9yBItRvTFELxo/view?usp=sharing' 
path = 'https://drive.google.com/uc?export=download&id='+url.split('/')[-2]
geodata = pd.read_csv(path, encoding = 'CP1252') # change encoding to 'mbcs' in Windows

On the "frame" dataset we have the ratings users have given to places. Ratings go from 0 to 2.

In [3]:
frame.head(3)

userID  placeID  rating  food_rating  service_rating
0  U1077   135085       2            2               2
1  U1077   135038       2            2               1
2  U1077   132825       2            2               2

In the `geodata` dataset we have info about the places. We will only use the `name` column.

In [4]:
geodata.head(2)

placeID   latitude   longitude  \
0   134999  18.915421  -99.184871   
1   132825  22.147392 -100.983092   

                                      the_geom_meter             name  \
0  0101000020957F000088568DE356715AC138C0A525FC46...  Kiku Cuernavaca   
1  0101000020957F00001AD016568C4858C1243261274BA5...  puesto de tacos   

                                  address        city    state country fax  \
0                              Revolucion  Cuernavaca  Morelos  Mexico   ?   
1  esquina santos degollado y leon guzman      s.l.p.   s.l.p.  mexico   ?   

   ...            alcohol smoking_area dress_code     accessibility   price  \
0  ...  No_Alcohol_Served         none   informal  no_accessibility  medium   
1  ...  No_Alcohol_Served         none   informal        completely     low   

                     url Rambience franchise    area other_services  
0  kikucuernavaca.com.mx  familiar         f  closed           none  
1                      ?  familiar         f    open           none  

[2 rows x 21 columns]

In [5]:
places =  geodata[['placeID', 'name']]
places.head()

placeID                            name
0   134999                 Kiku Cuernavaca
1   132825                 puesto de tacos
2   135106      El Rincón de San Francisco
3   132667  little pizza Emilio Portes Gil
4   132613                   carnitas_mata

In the `cuisine` dataset we have the type of cuisine that restaurants offer.

In [6]:
cuisine.head(3)

placeID        Rcuisine
0   135110         Spanish
1   135109         Italian
2   135107  Latin_American

## Popularity/Quality based recommmender system

Let's group places by rating, and look at their average rating. This is an **explicit** rating given by users.

In [7]:
rating = pd.DataFrame(frame.groupby('placeID')['rating'].mean())
rating.sort_values("rating", ascending=False).head()

rating
placeID          
132955   2.000000
135034   2.000000
134986   2.000000
132922   1.833333
132755   1.800000

The top rated places have a perfect score of 2/2. But how many reviews do these places have?

In [8]:
frame.query("placeID==132955")

userID  placeID  rating  food_rating  service_rating
934   U1004   132955       2            2               2
960   U1061   132955       2            2               2
996   U1059   132955       2            1               2
1014  U1097   132955       2            2               1
1080  U1096   132955       2            2               2

Looks like only 5 people went to this place. Maybe they're just the owner's friends! Or maybe they're really top-quality places, but too niche to recommend to the masses.

We can also look at how many times each restaurant has received a rating. The ratings count is an **implicit** rating.

In [9]:
rating['rating_count'] = frame.groupby('placeID')['rating'].count()
rating.sort_values("rating_count", ascending=False).head()

rating  rating_count
placeID                        
135085   1.333333            36
132825   1.281250            32
135032   1.178571            28
135052   1.280000            25
132834   1.000000            25

Some places have been visited around 30 times. They are more popular than the top rated places, but received lower explicit ratings.

Let's locate the most popular place, and get some info about it:

In [10]:
rating.sort_values('rating_count', ascending=False).head()

rating  rating_count
placeID                        
135085   1.333333            36
132825   1.281250            32
135032   1.178571            28
135052   1.280000            25
132834   1.000000            25

In [11]:
# placeId of most popular place
top_popular_placeID = rating.sort_values('rating_count', ascending=False).head(1).index[0]

# name of the most popular place
places[places['placeID']==top_popular_placeID]

placeID                    name
121   135085  Tortas Locas Hipocampo

In [12]:
# cuisine of the most popular place
cuisine[cuisine['placeID']==top_popular_placeID]

placeID   Rcuisine
44   135085  Fast_Food

The most popular place is "Tortas Locas Hipocampo", a fast food place that has received 36 reviews and it has an average score of 1.33.

### Challenge:

Find a hybrid system to sort restaurants, so that you can recommend the "best" places: restaurants that are both high rated and popular.

In [34]:
frame.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1161 entries, 0 to 1160
Data columns (total 5 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   userID          1161 non-null   object
 1   placeID         1161 non-null   int64 
 2   rating          1161 non-null   int64 
 3   food_rating     1161 non-null   int64 
 4   service_rating  1161 non-null   int64 
dtypes: int64(4), object(1)
memory usage: 45.5+ KB


In [35]:
frame.isna().sum()

userID            0
placeID           0
rating            0
food_rating       0
service_rating    0
dtype: int64

In [39]:
mean_rating_value = rating['rating'].mean()
mean_rating_value

1.1796215584942962

In [41]:
new = pd.merge(rating, places, on='placeID')
new.head()

placeID  rating  rating_count                name
0   132560    0.50             4  puesto de gorditas
1   132561    0.75             4          cafe ambar
2   132564    1.25             4             churchs
3   132572    1.00            15        Cafe Chaires
4   132583    1.00             4    McDonalds Centro

In [46]:
new.loc[(new['rating']>= 1.179) & (new['rating_count']>= 20)]

placeID    rating  rating_count                           name
31    132825  1.281250            32                puesto de tacos
84    135038  1.208333            24          Restaurant la Chalita
88    135042  1.250000            20    Restaurant Oriental Express
98    135052  1.280000            25         La Cantina Restaurante
106   135062  1.238095            21  Restaurante El Cielo Potosino
123   135085  1.333333            36         Tortas Locas Hipocampo

In [47]:
new1 = pd.merge(frame, new, on='placeID')
new1.head()

userID  placeID  rating_x  food_rating  service_rating  rating_y  \
0  U1077   135085         2            2               2  1.333333   
1  U1108   135085         1            2               1  1.333333   
2  U1081   135085         1            2               1  1.333333   
3  U1056   135085         2            2               2  1.333333   
4  U1134   135085         2            1               2  1.333333   

   rating_count                    name  
0            36  Tortas Locas Hipocampo  
1            36  Tortas Locas Hipocampo  
2            36  Tortas Locas Hipocampo  
3            36  Tortas Locas Hipocampo  
4            36  Tortas Locas Hipocampo

In [52]:
new1.loc[(new1['rating_x']>= 1.179) & (new1['rating_count']>= 35)].head()

userID  placeID  rating_x  food_rating  service_rating  rating_y  \
0  U1077   135085         2            2               2  1.333333   
3  U1056   135085         2            2               2  1.333333   
4  U1134   135085         2            1               2  1.333333   
5  U1116   135085         2            2               2  1.333333   
6  U1132   135085         2            2               2  1.333333   

   rating_count                    name  
0            36  Tortas Locas Hipocampo  
3            36  Tortas Locas Hipocampo  
4            36  Tortas Locas Hipocampo  
5            36  Tortas Locas Hipocampo  
6            36  Tortas Locas Hipocampo